src/Core/Framework/Test/DataAbstractionLayer/Dbal/RepositoryIteratorTest.php

```diff
@@ -96,4 +100,99 @@ public function testFetchIdAutoIncrement(): void
         }
         static::assertEquals($totalFetchedIds, 3);
     }
+
+    public function testFetchNotObviousEmptyNextRequestAutoIncrement(): void
+    {
+        /** @var EntityRepository<ProductCollection> $productRepository */
+        $productRepository = $this->getContainer()->get('product.repository');
+        /** @var EventDispatcher $eventDispatcher */
+        $eventDispatcher = $this->getContainer()->get('event_dispatcher');
+
+        $context = Context::createDefaultContext();
+
+        $builder = new ProductBuilder(new IdsCollection(), 'product1');
+        $builder->price(1);
+        $productRepository->create([$builder->build()], $context);
+
+        $builder = new ProductBuilder(new IdsCollection(), 'product2');
+        $builder->price(1);
+        $productRepository->create([$builder->build()], $context);
+
+        $builder = new ProductBuilder(new IdsCollection(), 'product3');
+        $builder->price(1);
+        $productRepository->create([$builder->build()], $context);
+
+        $criteria = new Criteria();
+        $criteria->setTitle('test__product_iteration');
+        // 3 products will be fetched in pairs of 2
+        // The last response will obviously have 1 (less than 2) item and this already indicates end
+        // so this should not need an additional search
+        $criteria->setLimit(2);
+
+        $searchesCount = 0;
+        $eventDispatcher->addListener(EntitySearchedEvent::class, function (EntitySearchedEvent $event) use (&$searchesCount): void {
```

Comment:
this is evil as nobody will remove your `addListener`. use `$this->addEventListener` that does that for you